pkg/generate/app/cmd/newapp.go

```diff
@@ -717,12 +717,73 @@ func (c *AppConfig) checkCircularReferences(objects app.Objects) error {
 	for _, obj := range objects {
 		if bc, ok := obj.(*buildapi.BuildConfig); ok {
 			input := buildutil.GetInputReference(bc.Spec.Strategy)
-			if bc.Spec.Output.To != nil && input != nil &&
-				reflect.DeepEqual(input, bc.Spec.Output.To) {
-				ns := input.Namespace
-				if len(ns) == 0 {
-					ns = c.OriginNamespace
+			output := bc.Spec.Output.To
+
+			if output == nil || input == nil {
+				return nil
+			}
+
+			ns := input.Namespace
+			if len(ns) == 0 {
+				ns = c.OriginNamespace
+			}
+
+			// We might be outputing a docker image that would trigger our imagestream repo.
+			if output.Kind == "DockerImage" && input.Kind == "ImageStreamTag" {
+				isName, isTag, ok := imageapi.SplitImageStreamTag(input.Name)
+				if !ok {
+					return fmt.Errorf("Unable to parse ImageStreamTag name: %q", input.Name)
 				}
+
+				// The imagestream is usually being created alongside the buildconfig
+				// when new-build is being used, so scan objects being created for it.
+				var is *imageapi.ImageStream
+				for _, check := range objects {
+					if is2, ok := check.(*imageapi.ImageStream); ok {
+						if is2.Name == isName {
+							is = is2
+							break
+						}
+					}
+				}
+
+				if is == nil {
+					var err error
+					is, err = c.OSClient.ImageStreams(ns).Get(isName)
+					if err != nil {
+						glog.Warningf("Unable to check for circular build input/outputs")
+						return nil
+					}
+				}
+
+				// Dereference ImageStream so that we compare DockerImage against DockerImage.
+				input = is.Spec.Tags[isTag].From
+			}
+
+			// For DeepEqual to work for all input variations, we need to massage the
+			// data a bit. Make copies so we don't affect actual data structures.
+			inCopy := *input
+			outCopy := *output
+
+			for _, ref := range []*kapi.ObjectReference{&inCopy, &outCopy} {
+				if ref.Kind == "DockerImage" {
+					// DockerImage names may or may not have a tag suffix. Add
+					// :latest if there is no tag so that string comparison will
+					// behave as expected.
+					if strings.Contains(ref.Name, ":") == false {
+						ref.Name += ":latest"
+					}
+				}
+				// Some code paths add namespace and others don't. Make things
+				// consistent.
```

Comment:
Not entirely sure, but I think the "inconsistency" is in the fact that we rely on the server to set the namespace.
For instance, with `oc create` one can create a pod, and the JSON spec doesn't set the namespace, but it will be set by the server upon creation.

@bparees does that make sense? I think not setting the namespace on the client side is intentional.